packages/datadog-plugin-google-cloud-pubsub/src/producer.js

```diff
@@ -2,42 +2,180 @@
+      this.tracer.inject(activeSpan, 'text_map', attributes)
+
+      // Add 128-bit trace ID upper bits if present
+      const traceIdUpperBits = activeSpan.context()._trace.tags['_dd.p.tid']
+      if (traceIdUpperBits) attributes['_dd.p.tid'] = traceIdUpperBits
+
+      // Add metadata
+      if (pubsub) attributes['gcloud.project_id'] = pubsub.projectId
+      if (topicName) attributes['pubsub.topic'] = topicName
+    } catch (err) {
+      // Silently fail - trace context injection is best-effort
+      log.warn(`Error injecting trace context into message attributes: ${err.message}`)
+    }
+  }
+
   bindStart (ctx) {
     const { request, api, projectId } = ctx
-
     if (api !== 'publish') return
 
     const messages = request.messages || []
     const topic = request.topic
-    const span = this.startSpan({ // TODO: rename
-      resource: `${api} ${topic}`,
+    const messageCount = messages.length
+    const hasTraceContext = messages[0]?.attributes?.['x-datadog-trace-id']
+
+    // Collect span links from messages 2-N (skip first - it becomes parent)
+    const spanLinkData = hasTraceContext
+      ? messages.slice(1).map(msg => this.#extractSpanLink(msg.attributes)).filter(Boolean)
+      : []
+
+    const firstAttrs = messages[0]?.attributes
+    const parentData = firstAttrs?.['x-datadog-trace-id'] && firstAttrs['x-datadog-parent-id']
+      ? {
+          traceId: firstAttrs['x-datadog-trace-id'],
+          spanId: firstAttrs['x-datadog-parent-id'],
+          traceIdUpper: firstAttrs['_dd.p.tid'],
+          samplingPriority: firstAttrs['x-datadog-sampling-priority']
+        }
+      : null
+
+    const topicName = topic.split('/').pop() || topic
+    const batchSpan = this.startSpan({
+      childOf: parentData ? this.#extractParentContext(parentData) : undefined,
+      resource: `${api} to Topic ${topicName}`,
       meta: {
         'gcloud.project_id': projectId,
-        'pubsub.method': api, // TODO: remove
-        'pubsub.topic': topic
+        'pubsub.method': api,
+        'pubsub.topic': topic,
+        'span.kind': 'producer',
+        '_dd.base_service': this.tracer._service,
+        '_dd.serviceoverride.type': 'integration',
+        'pubsub.linked_message_count': spanLinkData.length || undefined,
+        operation: messageCount > 1 ? 'batched.pubsub.request' : 'pubsub.request'
+      },
+      metrics: {
+        'pubsub.batch.message_count': messageCount,
+        'pubsub.batch': messageCount > 1 ? true : undefined
       }
     }, ctx)
 
-    for (const msg of messages) {
-      if (!msg.attributes) {
-        msg.attributes = {}
+    const spanCtx = batchSpan.context()
+    const batchTraceId = spanCtx.toTraceId()
+    const batchSpanId = spanCtx.toSpanId()
+    const batchTraceIdUpper = spanCtx._trace.tags['_dd.p.tid']
+    const batchTraceIdHex = BigInt(batchTraceId).toString(16).padStart(16, '0')
+    const batchSpanIdHex = BigInt(batchSpanId).toString(16).padStart(16, '0')
```

Comment:
```suggestion
    const batchTraceIdHex = spanCtx.toTraceId(true)
    const batchSpanIdHex = spanCtx.toSpanId(true)
    const batchTraceIdUpper = spanCtx._trace.tags['_dd.p.tid']
```

We have options to directly get the 128bits :)